File: src/domain/projects.ts
import { matchesText } from './categorize'
import { withEstimatedKinds } from './estimatedKinds'
import { computePeriodTotals } from './ledger'
import { addMoney, type Halalas } from './money'
import { normalizeText } from './normalize'
import type { Id, PersonAllocation, RuleMatchMode, Transaction } from './entities/types'
import type { Project, ProjectLink, ProjectRule, ProjectRuleDirection } from './entities/projectEntities'

/**
 * المشاريع — OVERRIDES §34. كل الحسابات هنا نقية؛ التخزين في `manageProjects`.
 * - «صرفت» = **نصيبك** (نفس مصروف الرئيسية) + المستبعد من الميزانية؛ «جالك» = الدخل. النوع اللي لسه
 *   ما اتأكدش بيتحسب تقريبي زي الرئيسية (`withEstimatedKinds`)، وكل عملية مرة واحدة.
 * - العملية ممكن تبقى في أكتر من مشروع.
 * - القاعدة بتضيف لوحدها العمليات اللي **اتسجلت بعدها** بس؛ القديمة بسؤال وقت عملها.
 */

export const PROJECT_NAME_MAX = 60
export const PROJECT_RULE_TEXT_MAX = 60

export class ProjectError extends Error {}

export function checkProjectName(name: string, projects: readonly Project[], selfId?: Id): { name: string; normalizedName: string } {
  const clean = name.trim().replace(/\s+/g, ' ')
  if (!clean || clean.length > PROJECT_NAME_MAX) throw new ProjectError(`اكتب اسم المشروع بحد أقصى ${PROJECT_NAME_MAX} حرف.`)
  const normalizedName = normalizeText(clean)
  if (projects.some((p) => p.id !== selfId && p.normalizedName === normalizedName)) throw new ProjectError('فيه مشروع بنفس الاسم.')
  return { name: clean, normalizedName }
}

const MODES: readonly RuleMatchMode[] = ['contains', 'startsWith', 'exact']
const DIRECTIONS: readonly ProjectRuleDirection[] = ['out', 'in', 'any']

export function checkProjectRule(input: { matchText: string; matchMode: RuleMatchMode; direction: ProjectRuleDirection }) {
  const matchText = input.matchText.trim().replace(/\s+/g, ' ')
  if (matchText.length < 2 || matchText.length > PROJECT_RULE_TEXT_MAX) {
    throw new ProjectError(`اكتب نص القاعدة من حرفين لـ ${PROJECT_RULE_TEXT_MAX} حرف.`)
  }
  if (!MODES.includes(input.matchMode)) throw new ProjectError('طريقة المطابقة مش معروفة.')
  if (!DIRECTIONS.includes(input.direction)) throw new ProjectError('الاتجاه مش معروف.')
  return { matchText, matchMode: input.matchMode, direction: input.direction }
}

/** معرّف ثابت من المشروع والعملية ⇒ ربط واحد بس لنفس العملية في نفس المشروع. */
export const projectLinkId = (projectId: Id, transactionId: Id) => `plink-${projectId}-${transactionId}`

/** العمليات اللي في المشروع فعلًا (المستبعدة لأ). */
export function memberIds(links: readonly ProjectLink[], projectId: Id): Set<Id> {
  return new Set(links.filter((l) => l.projectId === projectId && l.source !== 'excluded').map((l) => l.transactionId))
}

export function ruleMatchesTransaction(rule: Pick<ProjectRule, 'matchText' | 'matchMode' | 'direction'>, t: Transaction): boolean {
  if (rule.direction !== 'any' && t.observedDirection !== rule.direction) return false
  return [t.rawMerchantName, t.rawDescription].some((text) => !!text && matchesText(rule.matchText, rule.matchMode, text))
}

/** العمليات المطابقة للقاعدة ومش مربوطة بمشروعها بأي شكل (المستبعدة بإيد المستخدم ما بتتلمسش). */
export function ruleCandidates(rule: ProjectRule, transactions: readonly Transaction[], links: readonly ProjectLink[]): Transaction[] {
  const known = new Set(links.filter((l) => l.projectId === rule.projectId).map((l) => l.transactionId))
  return transactions.filter((t) => !known.has(t.id) && ruleMatchesTransaction(rule, t))
}

function ruleLink(rule: ProjectRule, t: Transaction, now: string): ProjectLink {
  return { id: projectLinkId(rule.projectId, t.id), projectId: rule.projectId, transactionId: t.id, source: 'rule', createdAt: now }
}

export function planRuleLinks(rule: ProjectRule, transactions: readonly Transaction[], links: readonly ProjectLink[], now: string): ProjectLink[] {
  return ruleCandidates(rule, transactions, links).map((t) => ruleLink(rule, t, now))
}

/**
 * العمليات الجديدة: كل قاعدة شغالة بتاخد العمليات اللي **اتسجلت بعدها** بس (القديمة اتسألت عنها وقت عملها).
 * قاعدتين في نفس المشروع على نفس العملية ⇒ ربط واحد.
 */
export function planSyncLinks(rules: readonly ProjectRule[], transactions: readonly Transaction[], links: readonly ProjectLink[], now: string): ProjectLink[] {
  const planned: ProjectLink[] = []
  for (const rule of rules) {
    if (!rule.enabled) continue
    const fresh = transactions.filter((t) => t.createdAt > rule.createdAt)
    planned.push(...planRuleLinks(rule, fresh, [...links, ...planned], now))
  }
  return planned
}

/** من إمتى نراجع العمليات الجديدة على الجهاز ده: آخر مراجعة، ومش قبل أقدم قاعدة شغالة. `null` = مفيش قواعد. */
export function syncStart(rules: readonly ProjectRule[], cursor: string | null): string | null {
  const oldest = rules.filter((r) => r.enabled).map((r) => r.createdAt).sort()[0]
  if (oldest === undefined) return null
  return cursor && cursor > oldest ? cursor : oldest
}

/**
 * إضافة أو شيل عملية من مشروع بإيد المستخدم — الروابط اللي هتتحفظ (فاضية = مفيش تغيير).
 * الشيل بيسيبها «مستبعدة» عشان قاعدة ما ترجعهاش. `existing` = روابط العملية دي في المشروع ده؛ ممكن تبقى
 * أكتر من واحد بمعرّف غير المعتاد (استعادة نسخة دمجت العملية في عملية موجودة بمعرّف تاني).
 */
export function membershipChanges(existing: readonly ProjectLink[], projectId: Id, transactionId: Id, member: boolean, now: string): ProjectLink[] {
  if (!member) return existing.filter((l) => l.source !== 'excluded').map((l) => ({ ...l, source: 'excluded' }))
  if (existing.some((l) => l.source !== 'excluded')) return []
  const first = existing[0]
  return [{ id: first?.id ?? projectLinkId(projectId, transactionId), projectId, transactionId, source: 'manual', createdAt: first?.createdAt ?? now }]
}

export interface ProjectSummary {
  spentMinor: Halalas
  receivedMinor: Halalas
  count: number
  /** كام عملية نوعها لسه مش مؤكد فالرقم «تقريبي» (زي الرئيسية). */
  estimatedCount: number
  /** منهم كام محتاجة تأكيد. */
  needsReviewCount: number
}

export function summarizeProject(
  transactions: readonly Transaction[],
  allocations: readonly PersonAllocation[],
  categoryNameById: ReadonlyMap<string, string>,
): ProjectSummary {
  const view = withEstimatedKinds(transactions, categoryNameById)
  const totals = computePeriodTotals(view.transactions, allocations)
  return {
    spentMinor: addMoney(totals.personalExpenseMinor, totals.excludedExpenseMinor),
    receivedMinor: totals.incomeMinor,
    count: transactions.length,
    estimatedCount: view.estimatedCount,
    needsReviewCount: view.needsReviewCount,
  }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Projects.kt
package app.masroufy.core

/**
 * المشاريع — نقل `src/domain/projects.ts` + الكيانات (OVERRIDES §34). **علامة إضافية مش تصنيف**.
 * «صرفت» = نصيبك + المستبعد من الميزانية؛ «جالك» = الدخل؛ غير المؤكد «تقريبي» زي الرئيسية.
 * القاعدة بتضيف لوحدها اللي **اتسجل بعدها** بس؛ القديم بسؤال وقت عملها.
 */
data class Project(val id: Id, val name: String, val normalizedName: String, val archived: Boolean, val createdAt: String)

/** "manual" / "rule" / "excluded" (المستخدم شالها — بتفضل عشان قاعدة ما ترجعهاش). */
data class ProjectLink(val id: Id, val projectId: Id, val transactionId: Id, val source: String, val createdAt: String)

data class ProjectRule(
    val id: Id,
    val projectId: Id,
    val matchText: String,
    val matchMode: RuleMatchMode,
    /** "out" / "in" / "any". */
    val direction: String,
    val enabled: Boolean,
    val createdAt: String,
)

const val PROJECT_NAME_MAX = 60
const val PROJECT_RULE_TEXT_MAX = 60

class ProjectError(message: String) : IllegalArgumentException(message)

data class CheckedName(val name: String, val normalizedName: String)

fun checkProjectName(name: String, projects: List<Project>, selfId: Id? = null): CheckedName {
    val clean = JsText.collapseWhitespace(JsText.trim(name))
    if (clean.isEmpty() || clean.length > PROJECT_NAME_MAX) throw ProjectError("اكتب اسم المشروع بحد أقصى $PROJECT_NAME_MAX حرف.")
    val normalized = normalizeText(clean)
    if (projects.any { it.id != selfId && it.normalizedName == normalized }) throw ProjectError("فيه مشروع بنفس الاسم.")
    return CheckedName(clean, normalized)
}

data class CheckedRule(val matchText: String, val matchMode: RuleMatchMode, val direction: String)

fun checkProjectRule(matchText: String, matchMode: RuleMatchMode, direction: String): CheckedRule {
    val text = JsText.collapseWhitespace(JsText.trim(matchText))
    if (text.length < 2 || text.length > PROJECT_RULE_TEXT_MAX) throw ProjectError("اكتب نص القاعدة من حرفين لـ $PROJECT_RULE_TEXT_MAX حرف.")
    if (direction !in listOf("out", "in", "any")) throw ProjectError("الاتجاه مش معروف.")
    return CheckedRule(text, matchMode, direction)
}

/** معرّف ثابت ⇒ ربط واحد بس لنفس العملية في نفس المشروع. */
fun projectLinkId(projectId: Id, transactionId: Id) = "plink-$projectId-$transactionId"

/** العمليات اللي في المشروع فعلًا (المستبعدة لأ). */
fun memberIds(links: List<ProjectLink>, projectId: Id): Set<Id> =
    links.filter { it.projectId == projectId && it.source != "excluded" }.map { it.transactionId }.toCollection(LinkedHashSet())

fun ruleMatchesTransaction(matchText: String, matchMode: RuleMatchMode, direction: String, t: Transaction): Boolean {
    if (direction != "any" && t.observedDirection.wire != direction) return false
    return listOf(t.rawMerchantName, t.rawDescription).any { !it.isNullOrEmpty() && matchesText(matchText, matchMode, it) }
}

/** المطابقة ومش مربوطة بمشروعها بأي شكل (المستبعدة بإيد المستخدم ما بتتلمسش). */
fun ruleCandidates(rule: ProjectRule, transactions: List<Transaction>, links: List<ProjectLink>): List<Transaction> {
    val known = links.filter { it.projectId == rule.projectId }.map { it.transactionId }.toSet()
    return transactions.filter { it.id !in known && ruleMatchesTransaction(rule.matchText, rule.matchMode, rule.direction, it) }
}

fun planRuleLinks(rule: ProjectRule, transactions: List<Transaction>, links: List<ProjectLink>, now: String): List<ProjectLink> =
    ruleCandidates(rule, transactions, links).map { ProjectLink(projectLinkId(rule.projectId, it.id), rule.projectId, it.id, "rule", now) }

/** كل قاعدة شغالة بتاخد اللي **اتسجل بعدها** بس؛ قاعدتين على نفس العملية في نفس المشروع ⇒ ربط واحد. */
fun planSyncLinks(rules: List<ProjectRule>, transactions: List<Transaction>, links: List<ProjectLink>, now: String): List<ProjectLink> {
    val planned = mutableListOf<ProjectLink>()
    for (rule in rules) {
        if (!rule.enabled) continue
        planned += planRuleLinks(rule, transactions.filter { it.createdAt > rule.createdAt }, links + planned, now)
    }
    return planned
}

/** من إمتى نراجع الجديد: آخر مراجعة، ومش قبل أقدم قاعدة شغالة. null = مفيش قواعد. */
fun syncStart(rules: List<ProjectRule>, cursor: String?): String? {
    val oldest = rules.filter { it.enabled }.map { it.createdAt }.minOrNull() ?: return null
    return if (!cursor.isNullOrEmpty() && cursor > oldest) cursor else oldest
}

/** إضافة أو شيل بإيد المستخدم — الروابط اللي هتتحفظ (فاضية = مفيش تغيير). الشيل بيسيبها «مستبعدة». */
fun membershipChanges(existing: List<ProjectLink>, projectId: Id, transactionId: Id, member: Boolean, now: String): List<ProjectLink> {
    if (!member) return existing.filter { it.source != "excluded" }.map { it.copy(source = "excluded") }
    if (existing.any { it.source != "excluded" }) return emptyList()
    val first = existing.firstOrNull()
    return listOf(ProjectLink(first?.id ?: projectLinkId(projectId, transactionId), projectId, transactionId, "manual", first?.createdAt ?: now))
}

data class ProjectSummary(val spentMinor: Halalas, val receivedMinor: Halalas, val count: Int, val estimatedCount: Int, val needsReviewCount: Int)

fun summarizeProject(transactions: List<Transaction>, allocations: List<PersonAllocation>, categoryNameById: Map<String, String>): ProjectSummary {
    val view = withEstimatedKinds(transactions, categoryNameById)
    val totals = computePeriodTotals(view.transactions, allocations)
    return ProjectSummary(addMoney(totals.personalExpenseMinor, totals.excludedExpenseMinor), totals.incomeMinor, transactions.size, view.estimatedCount, view.needsReviewCount)
}
